### ISSUE TITLE
Where is documentation on the PowerShell language?

### ISSUE THREAD

**Issue by RossBoylan** [MicrosoftDocs/PowerShell-Docs#4020]
I've been round and round your site looking for the language specification.
It isn't even mentioned in the "how to learn powershell" blog page.  I don't want to watch some videos, which is mostly what that page references.
I would think it would be in the PowerShell reference.  But it's not that I can see.
Information on how to debug would be useful too.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: f1ee277d-73d7-7abc-4928-fd88d097ed55
* Version Independent ID: 1d4271db-8317-37ad-4034-1e7f79157532
* Content: [Windows PowerShell Reference](https://docs.microsoft.com/en-us/powershell/developer/windows-powershell-reference#feedback)
* Content Source: [developer/windows-powershell-reference.md](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/developer/windows-powershell-reference.md)
* Product: **powershell**
* GitHub Login: @joeyaiello
* Microsoft Alias: **jaiello**

**Comment by sdwheeler** [comment]
Most of what you are looking for is in the About_* topics of the PowerShell user documentation. This article is in the PowerShell SDK documentation, which is about extending PowerShell or creating other host applications.

Start here: https://docs.microsoft.com/en-us/powershell/scripting/learn/understanding-important-powershell-concepts?view=powershell-6

Then check out the About_* topics  here: https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/?view=powershell-6

**Comment by RossBoylan** [comment]
I don't think this issue is at all resolved.  It was a complaint that there is no obvious path from the high-level/introductory materials to the language definition.  There still isn't.

Second, the about_* topics are not a substitute for an actual language specification, any more than having a bunch of examples is.

I wanted something like https://www.microsoft.com/en-us/download/details.aspx?id=36389, a link I only found in a response on stackoverflow.  However, that's for an older version, and other chatter indicates MS has not produced an updated version (just "what's new in ..." documents) and apparently doesn't think it should.  Because they don't want people to use PowerShell?

**Comment by sdwheeler** [comment]
That download you found is the only published version of the language specification. The language hasn't changed since then. Only the functionality of the built-in cmdlets has changed over the new versions.

I agree that the docs are lacking a coherent learning path for the syntax of the language. This is something we are working to improve. For now, the About_* topics are the best we have.

**Comment by doctordns** [comment]
It might be an idea to take that download and put it up on docs.microsoft.com.

And I think Sean is not entirely correct about the language. PowerShell 6 IS changing things and there _may_ be some small updates needed. 

But whatever - putting it into docs.microsoft.com seems a good idea. This is not the first time this issue has come up! :-)

**Comment by joeyaiello** [comment]
@RossBoylan: just want to make sure you've seen our ["About" topics](https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/). If there's any aspect of the languages that's not covered in those, you should file issues to make sure we get them fixed. 

As for a formal "language specification": I wasn't around when we originally did the 3.0 spec, but my understanding is that it was a *significant* amount of work and that the value provided to the community was comparably very low. If you don't feel like this is accurate, it'd be helpful to understand what you see as the functional delta between the About topics and what you see as an ideal specification.

**Comment by RossBoylan** [comment]
This started when I wanted to find how to concatenate strings and  what the string syntax was, e.g., escape characters and handling of line continuation.  I did spend some time looking at the about topics.  BTW, the link to the About topics in the previous post doesn't quite work.

Looking now, I look for topics on Syntax, Strings, Expressions--nope.  About Objects exists, but nothing seems relevant to Strings.  I'm still not sure if "everything is an object," but I'd guess not.  Operators does document + and says it concatenates elements, which literally doesn't seem relevant (sounds as if it's for lists), though Arithmetic Operators does say + does string concatenation, and also suggests some of the hazards of attempting to use it that way.  Having discovered -join in the spec, I now see there is an about topic for it.  The discussion of string concatenation under "Arithmetic Operators" is unfortunate because in that context + is a "String Operator".

So the about stuff has 2 limitations: there's stuff that's not there, and finding what you want is a treasure hunt.  I was frankly stunned that such information was not more readily available.  Are we supposed to learn the language by reading through the topics alphabetically?

It would certainly be useful to have something that was friendlier than the formal language specification.  For example, having discovered the formal specification of the string syntax in 2.3.5.2 of the 3.0 language specification, I must confess the meaning is not immediately apparent :)  The documentation of Python is a very good model, including the fact that it has an overview/intro to the language (https://docs.python.org/3/tutorial/index.html) and a language reference (https://docs.python.org/3/reference/index.html).  That reference (2.4.1 https://docs.python.org/3/reference/lexical_analysis.html#literals)  has a BNF for string literals giving the formal syntax but then, one paragraph below it, "In plain English: Both types of literals can be enclosed in matching single quotes (') or double quotes ("). They can also be enclosed in matching groups of three single or double quotes ...."

Sometime later I wondered if commands (or cmdlets) executed synchronously or asynchronously; I couldn't find any explicit statements, and concluded it would be too crazy if they were asynch (also the search on asynchronous turned up just a few hits, suggesting it was a special case).

**Comment by RossBoylan** [comment]
BTW, I'm a little puzzled about whom the intended audience for PowerShell is.  A lot of the discussion suggests it's for System Administrators, but that doesn't seem like a group that's going to get into using .NET frameworks.  And programmers already have a lot languages to choose from, so why do they need another?

**Comment by doctordns** [comment]
Some comments to the most recent comment:
The intended audience for PowerShell is the same as it ever was: the IT Professional working in an IT environment using Windows (and more recently Linux and Mac). Systems admins, as you describe them, are very much a target audience. 

PowerShell is a .NET application that makes heavy use of the .NET BCL. If you use, for example, the Get-Process cmdlet to return running processes on a server, that cmdlet simply used the GetProesses() method of the .NET class System.Diagnostics.Process. IMHO, knowing more about the environment just makes you better (and worthy of a higher salary).  

I teach this subject, and (broadly) see three types of admins:

1. The junior engineer, guided by Google who just wants to get something done. For example, Exchange's self-signed certificates have lapsed and BillyBob just wants to fix it. A quick Google search brings up the relevant incantation  - job done. I agree that class of admin, there probably is no need to know much about .NET. 

2. The accomplished Admin - someone who is using, writing, and updating scripts to automate operations. In many cases, this class of admin is well served by the cmdlets on offer. SueBob can use the AD cmdlets that take an HR produced CSV file of new employees and ensure the users are added to the organisation's AD. This class of user may benefit from being able to leverage things from .NET. There may be .NET classes, values, and methods that could be useful for less common operations. Personally, I see a lot of these folks and they do benefit from knowing HOW PowerShell does what it does.

3. The Professional admin and/ the toolmaker. This class of admin understands the .NET Framework's CLR, 
knows how to navigate through and take full advantage of the thousands of classes in .NET's Base Class Libraries to solve problems and build great tools for less skilled folks to use. They understand the .NET limitations (eg how strings are stored on the stack, and what happens when you do **$String=$String+''1"**. JoeSueBob could also write simple forms based mini GUIs for junior admins and end users to use to solve common administrative problems. 
 
So sure, some user's don't need to know .NET, but any more experienced admin is going to want to know the nuts and bolts. And the folks that do understand this stuff, in my experience, are better compensated. 

As to why PowerShell? Well, great question. To some degree, you have to look back to 2001/2 to when PowerShell was just being conceived. At that time, the issue was two-fold: first GUIs do not scale (and still don't). To move NT from a departmental server to an enterprise server meant having to scale to thousands of servers. The GUI just does not scale. The second issue was that the Unix/Linux guys were right - scripting does scale - but the scripting languages available simply were not good enough. BASH sounds a great idea - but it's not object orientated, it's text-based. Which is great for Linux which is configured and managed using text files. But Windows is API orientated - and those older tools did not do a great job. With the product teams producing thousands of task-oriented cmdlets, using another language meant re-inventing the wheel. Given the massive amount of documentation (and books, web sites, training courses), if you want to be a great admin in the Windows Space, PowerShell gives you so much, going elsewhere may be pointless.

**Comment by doctordns** [comment]
Regarding your earlier query -, yes, in PowerShell - everything is a object:
42 - this is an object of the class System.Int32 which has a number of methods you can call to manipulate the integer
"Have a nice day" - this is an object of the class System.String which also has a nummber of methods you can call to manipulate that integer.
etc.
If you pipe anything to Get-Member you can see what that things is.

**Comment by doctordns** [comment]
> @RossBoylan: just want to make sure you've seen our ["About" topics](https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/). If there's any aspect of the languages that's not covered in those, you should file issues to make sure we get them fixed.

@joeyaiello for some reason th at link is broken. If you navigate to the 1st file in the list(https://docs.microsoft.com/powershell/module/microsoft.powershell.core/about/about_alias_provider?view=powershell-6 then click the link just above it in the tree control on the left hand side of the page, you go to a 404. There is a bug here.

**Comment by sdwheeler** [comment]
@RossBoylan @doctordns The link to the About topics is being fixed. It appears there is a new bug in our incremental build system that is breaking those links so I have to force a full build.

**Comment by RossBoylan** [comment]
BTW, maybe related to the build system problems, over the last few days the syntax summaries in the power shell documents seem to get repeated over and over.  E.g., https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/import-module?view=powershell-6
In that case I think each invocation is slightly different, but I think in others they've been identical.  But even when they are slightly different, is that the way you want to show them?  I don't remember noticing this repetition earlier.

Thanks for the other info.

Ross

________________________________________

**Comment by sdwheeler** [comment]
The syntax summaries are not repeated. There are multiple parameter sets. Think of parameter sets as method overrides. Each parameter set has an different name. Unfortunately, the names are not rendered on the web page but they are included in the Markdown source on GitHub.

You can see the names of the parameter sets using the following commands:

```powershell
(Get-Command Import-Module | Select-Object -ExpandProperty ParameterSets).name
```

**Comment by JohnL4** [comment]
Adding my grumpiness to this. I found the version 3.0 doc above through this github issue, after a google search.

This web page [ https://docs.microsoft.com/en-us/powershell/scripting/developer/prog-guide/windows-powershell-programmer-s-guide?view=powershell-6 ] says «For complete scripting information, see the Windows PowerShell Language Reference shipped with Windows PowerShell.» Where the heck is that? (I'm using PS 5.1)

I'm trying to find out about the "enum" keyword (which doesn't seem to appear in that doc). Maybe other folks are happy googling and copying sample code and modifying it until it does what they want, but I'm a first-principles sort of developer and I expect better. I don't need a formal spec because I'm not re-implementing PowerShell, but little snippets of almost-relevant info scattered about are not satisfying.

Is Dr. Scripto really supposed to be the primary source of reference info in conjunction with the about_* topics?

>:(

**Comment by Liturgist** [comment]
Links are dead.

> I've been round and round your site looking for the language specification.
> It isn't even mentioned in the "how to learn powershell" blog page. I don't want to watch some videos, which is mostly what that page references.
> I would think it would be in the PowerShell reference. But it's not that I can see.
> Information on how to debug would be useful too.
> 
> #### Document Details
> ⚠ _Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking._
> 
> * ID: f1ee277d-73d7-7abc-4928-fd88d097ed55
> * Version Independent ID: 1d4271db-8317-37ad-4034-1e7f79157532
> * Content: [Windows PowerShell Reference](https://docs.microsoft.com/en-us/powershell/developer/windows-powershell-reference#feedback)
> * Content Source: [developer/windows-powershell-reference.md](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/developer/windows-powershell-reference.md)
> * Product: **powershell**
> * GitHub Login: @joeyaiello
> * Microsoft Alias: **jaiello**

**Comment by sdwheeler** [comment]
https://www.microsoft.com/en-us/download/details.aspx?id=36389

**Comment by Liturgist** [comment]
Thank you for the current link. Surely there are a few changes from 3.0 to 5.1.
What is the prospect for documentation on PowerShell 6.x, 7.x, ...?
Is there an Antlr grammar available?

**Comment by sdwheeler** [comment]
There are no plans to update that document. There is no Antlr grammar available.

**Comment by Liturgist** [comment]
That sounds like sub-optimal product governance. What kind of organization would do that?
